### PR TITLE
Update the epoch checkpoint with the proper position before going offline for truncation

### PR DESCRIPTION
**Pull request by shaan1337**
Fixed: Update the epoch checkpoint with the proper position before going offline for truncation

Fixes: #3286

### CODE REVIEW

File: src/EventStore.Core.Tests/Services/Storage/EpochManager/when_truncating_the_epoch_checkpoint.cs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+		}
+
+		[TestFixture(typeof(LogFormat.V2), typeof(string))]
+		[TestFixture(typeof(LogFormat.V3), typeof(uint))]
+		public class with_cache_size_plus_two_epochs : when_truncating_the_epoch_checkpoint<TLogFormat, TStreamId> {
+			public with_cache_size_plus_two_epochs() : base(CachedEpochCount + 2) { }
+
+			[Test]
+			public void cannot_truncate_before_third_epoch_position() {
+				Assert.False(_epochManager.TryTruncateBefore(_epochs[2].EpochPosition, out _));
```
[the remark this answers is not in the document]
Comment (shaan1337) [reply src/EventStore.Core.Tests/Services/Storage/EpochManager/when_truncating_the_epoch_checkpoint.cs:224]: @hayley-jean yes that's correct. there are 1000 epoch checkpoints cached by default - so in practice it should not happen.

File: src/EventStore.Core/Services/ClusterStorageWriterService.cs [hunk 2]
```diff
@@ -131,9 +131,37 @@ public class ClusterStorageWriterService<TStreamId> : StorageWriterService<TStre
...
+					message.SubscriptionPosition, message.SubscriptionPosition);
 				Db.Config.TruncateCheckpoint.Write(message.SubscriptionPosition);
 				Db.Config.TruncateCheckpoint.Flush();
 
+				// try to write the new epoch position prior to shutting down to
+				// avoid scanning the transaction log to find a valid epoch
+				// when starting up for truncation
+				var oldEpoch = EpochManager.GetLastEpoch();
+				if (EpochManager.TryTruncateBefore(message.SubscriptionPosition, out var newEpoch)) {
+					Log.Information("Truncated epoch from "
```
Comment (hayley-jean) [inline src/EventStore.Core/Services/ClusterStorageWriterService.cs:144]: Not a blocker to me, but if there are no newer epochs on the truncated node, you'll get a log that might be confusing (essentially you're truncating the epoch to the same position as it was before):

```
Truncated epoch from E134@1276418448:{66b39e7c-005b-4c44-be8b-fab3106a839b} to E134@1276418448:{66b39e7c-005b-4c44-be8b-fab3106a839b}
```

Perhaps the log could simply say `Truncated epoch to {newEpochNumber}@{newEpochPosition}:{newEpochId}` and conditionally append `from {oldEpochNumber}@{oldEpochPosition}:{oldEpochId}` if the old epoch was different to the new one? So in a case where the epochs are the same you'll get:

```
Truncated epoch to E134@1276418448:{66b39e7c-005b-4c44-be8b-fab3106a839b}
```

and in cases where they differ it'll be:

```
Truncated epoch to E134@1276418448:{66b39e7c-005b-4c44-be8b-fab3106a839b} from E136@1845951600:{c884fe93-27f7-4f30-9096-5718fff58265}
```

**Review by hayley-jean — APPROVED** [review]
Looks good and works locally 👍
I just have one comment about the log message we show when truncating the epoch.

**Review by hayley-jean — APPROVED** [review]

**Review by timothycoleman — COMMENTED** [review]
We are taking the two locks in different orders in different places, so there is a chance of a deadlock depending on how the EpochManager is used. Possibly we can get away with only one lock

**Comment by timothycoleman** [comment]
Apart from that looks good to me :+1:

**Review by timothycoleman — APPROVED** [review]
looks good :+1:

**Review by hayley-jean — APPROVED** [review]

**Pull request merged**